# R assignment
library(dplyr)
library(tidyverse)
library(stringr)
library(reshape2)
##this is delimited by tab
snp<-read.table("/Users/dan/Desktop/BCB546X-Spring2017/UNIX_Assignment/snp_position.txt", header = T,
sep="\t")
genotype<-read.table("/Users/dan/Desktop/BCB546X-Spring2017/UNIX_Assignment/fang_et_al_genotypes.txt",
sep="\t")
dim(snp)
dim(genotype)
##take the columns of interest, ie the SNP_ID, Chromosome, and position columns.
##same output, comma is to be clear (snp_cut_col<-snp[c(1,3,4)] snp_cut_col<-snp[,c(1,3,4)]
##alternatively, I can use:
snp_cut_col<-as.data.frame(snp[c(1,3,4)])
##taking out all the maize varieties and teosinte varieties into separate objects.
maize<-genotype %>% filter(V3=="ZMMIL"|V3=="ZMMLR"|V3=="ZMMMR")
teosinte<-genotype %>% filter(V3=="ZMPBA"|V3=="ZMPIL"|V3=="ZMPJA")
##since the new varieties object has no header, add
#header (SNP_ID column from genotype) back in as header so merging is possible later using this as common column.
id<-genotype[1,]
maize_header<-rbind(id,maize)
teosinte_header<-rbind(id, teosinte)
##then transpose maize and teosinte's rows for columns, into objects t_maize and t_teosinte.
t_maize_header<-as.data.frame(t(maize_header))
t_teosinte_header <- as.data.frame(t(teosinte_header))
##check if there are repeating row names
length(unique(t_maize_header))
length(t_maize_header)
length(unique(t_teosinte_header))
length(t_teosinte_header)
##next join the snp_cut_col to transposed maize or teosinte genotypes using SNP_ID as common column
colnames(t_maize_header)[1]<-"SNP_ID"
maize_join<-full_join(snp_cut_col,t_maize_header,by="SNP_ID")
colnames(t_teosinte_header)[1]<-"SNP_ID"
teosinte_join <- full_join(snp_cut_col,t_teosinte_header,by="SNP_ID")
##replace ?/? with ?? in the joined maize file, into new object maize_join_sub
maize_join_sub<-as.data.frame(lapply(maize_join, function(x) {
gsub("\\?\\/\\?", "\\?\\?",x)
}))

##for maize need 20 files, 10 for increasing order of chromosomes with ??
maize_join_sub_in<-maize_join_sub %>% arrange(Chromosome)

##make the first 10 files of maize with each chromosome per file.
maize_chrom01_incre<-subset(maize_join_sub_in, Chromosome=="1")
maize_chrom02_incre<-subset(maize_join_sub_in, Chromosome=="2")
maize_chrom03_incre<-subset(maize_join_sub_in, Chromosome=="3")
maize_chrom04_incre<-subset(maize_join_sub_in, Chromosome=="4")
maize_chrom05_incre<-subset(maize_join_sub_in, Chromosome=="5")
maize_chrom06_incre<-subset(maize_join_sub_in, Chromosome=="6")
maize_chrom07_incre<-subset(maize_join_sub_in, Chromosome=="7")
maize_chrom08_incre<-subset(maize_join_sub_in, Chromosome=="8")
maize_chrom09_incre<-subset(maize_join_sub_in, Chromosome=="9")
maize_chrom10_incre<-subset(maize_join_sub_in, Chromosome=="10")
##then replace the ?? in joined maize file with --,
maize_join_sub_de<-as.data.frame(lapply(maize_join, function(x) {
gsub("\\?\\/\\?", "\\-\\-",x)
}))
##rearrange the same object, in decreasing chromosome order, into new object maize_merge_sub, and replace ?? with --
maize_join_sub_de<-maize_join_sub %>% arrange(desc(Chromosome)) %>%
lapply(function(x) {
gsub("\\?\\?", "\\-\\-",x)
})
##the above will make a list, so then make this into dataframe.
maize_join_sub_de<-as.data.frame(maize_join_sub_de)
##make the second 10 files of maize with decreasing chromosome order, 1 chromosome per file
maize_chrom01_decre<-subset(maize_join_sub_de, Chromosome=="1")
maize_chrom02_decre<-subset(maize_join_sub_de, Chromosome=="2")
maize_chrom03_decre<-subset(maize_join_sub_de, Chromosome=="3")
maize_chrom04_decre<-subset(maize_join_sub_de, Chromosome=="4")
maize_chrom05_decre<-subset(maize_join_sub_de, Chromosome=="5")
maize_chrom06_decre<-subset(maize_join_sub_de, Chromosome=="6")
maize_chrom07_decre<-subset(maize_join_sub_de, Chromosome=="7")
maize_chrom08_decre<-subset(maize_join_sub_de, Chromosome=="8")
maize_chrom09_decre<-subset(maize_join_sub_de, Chromosome=="9")
maize_chrom10_decre<-subset(maize_join_sub_de, Chromosome=="10")
##for teosinte part
##replace ?/? with ?? in the joined teosinte file, into new object teosinte_join_sub
teosinte_join_sub<-as.data.frame(lapply(teosinte_join, function(x) {
gsub("\\?\\/\\?", "\\?\\?",x)
}))
##for teosinte need 20 files, 10 for increasing order of chromosomes
teosinte_join_sub_in<-teosinte_join_sub %>% arrange(Chromosome)
##make the first 10 files of teosinte with each chromosome per file.
teosinte_chrom01_incre<-subset(teosinte_join_sub_in, Chromosome=="1")
teosinte_chrom02_incre<-subset(teosinte_join_sub_in, Chromosome=="2")
teosinte_chrom03_incre<-subset(teosinte_join_sub_in, Chromosome=="3")
teosinte_chrom04_incre<-subset(teosinte_join_sub_in, Chromosome=="4")
teosinte_chrom05_incre<-subset(teosinte_join_sub_in, Chromosome=="5")
teosinte_chrom06_incre<-subset(teosinte_join_sub_in, Chromosome=="6")
teosinte_chrom07_incre<-subset(teosinte_join_sub_in, Chromosome=="7")
teosinte_chrom08_incre<-subset(teosinte_join_sub_in, Chromosome=="8")
teosinte_chrom09_incre<-subset(teosinte_join_sub_in, Chromosome=="9")
teosinte_chrom10_incre<-subset(teosinte_join_sub_in, Chromosome=="10")
##then replace the ?? in joined teosinte file with --,
teosinte_join_sub_de<-as.data.frame(lapply(teosinte_join, function(x) {
gsub("\\?\\/\\?", "\\-\\-",x)
}))
##rearrange the same object, in decreasing chromosome order, into new object teosinte_merge_sub, and replace ?? with --
teosinte_join_sub_de<-teosinte_join_sub %>% arrange(desc(Chromosome)) %>%
lapply(function(x) {
gsub("\\?\\?", "\\-\\-",x)
})
teosinte_join_sub_de<-as.data.frame(teosinte_join_sub_de)
##make the second 10 files of maize with decreasing chromosome order, 1 chromosome per file
teosinte_chrom01_decre<-subset(teosinte_join_sub_de, Chromosome=="1")
teosinte_chrom02_decre<-subset(teosinte_join_sub_de, Chromosome=="2")
teosinte_chrom03_decre<-subset(teosinte_join_sub_de, Chromosome=="3")
teosinte_chrom04_decre<-subset(teosinte_join_sub_de, Chromosome=="4")
teosinte_chrom05_decre<-subset(teosinte_join_sub_de, Chromosome=="5")
teosinte_chrom06_decre<-subset(teosinte_join_sub_de, Chromosome=="6")
teosinte_chrom07_decre<-subset(teosinte_join_sub_de, Chromosome=="7")
teosinte_chrom08_decre<-subset(teosinte_join_sub_de, Chromosome=="8")
teosinte_chrom09_decre<-subset(teosinte_join_sub_de, Chromosome=="9")
teosinte_chrom10_decre<-subset(teosinte_join_sub_de, Chromosome=="10")





##below is to make data tidy
##need to change headers and add in the SNP, chromosome, and position headers
colnames(maize_join_sub_in)<-as.character(unlist(maize_join_sub_in[984,]))
colnames(maize_join_sub_in)[1]<-"SNP_ID"
colnames(maize_join_sub_in)[2]<-"Chromosome"
colnames(maize_join_sub_in)[3]<-"Position"
##cut out the troublesome rows
##maize_join_sub_in<-maize_join_sub_in[-(984:986),]

maize_group<-maize_join_sub_in
colnames(maize_group)<-as.character(unlist(maize_group[986,]))
colnames(maize_group)[1]<-"SNP_ID"
colnames(maize_group)[2]<-"Chromosome"
colnames(maize_group)[3]<-"Position"

dim(maize_join_sub_in)
tidied_maize<-maize_join_sub_in %>% gather(key=Sample_ID, value=genotypes, 4:1576)
dim(maize_group)
tidied_group<-melt(maize_group, id=c("SNP_ID","Chromosome","Position"))
colnames(tidied_group)[4]<-"Group"
colnames(tidied_group)[5]<-"genotypes"

add_group<-full_join(tidied_maize, tidied_group, by=c("SNP_ID","Position","Chromosome","genotypes"))
##drop the last 3 rows
add_group<-add_group[-(3050451:3050453),]

ggplot(add_group)+geom_bar(aes(x=Chromosome))


##then if homozygous, then assign "homozygous" to the column homohetero
add_group<-within(add_group, {
  homohetero<-NA
  homohetero[genotypes=="A/A"|genotypes=="T/T"|genotypes=="C/C"|genotypes=="G/G"]<-"homozygous"
  homohetero[genotypes=="A/T"|genotypes=="A/C"|genotypes=="A/G"|genotypes=="C/T"|genotypes=="C/G"|genotypes=="C/A"|
               genotypes=="T/A"|genotypes=="T/C"|genotypes=="T/G"|genotypes=="G/A"|genotypes=="G/C"|genotypes=="G/T"]<-"heterozygous"
})

add_group<-as.data.frame(lapply(add_group, function(x) {
  gsub("\\?\\?", NA,x)
}))

add_group<-add_group %>% arrange(Sample_ID, Group)


ggplot(add_group,aes(x=Sample_ID, fill=homohetero))+
  geom_bar(position="fill")+
  facet_wrap(~Group)

ggplot(add_group)+geom_bar(aes(x=homohetero))+facet_wrap(~Group)
